src/KokkosPluginLoader.cpp

```diff
@@ -0,0 +1,136 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/util/KokkosPluginLoader.hpp"
+
+#include <dlfcn.h>
+#include <dirent.h>
+
+const uint64_t kokkos_interface_version = 20171029;
+
+inline
+bool
+isSharedObject(const std::string& filename)
+{
+  return (filename.size() > 3 && !filename.compare(filename.size() - 3, 3, ".so"));
+}
+
+namespace RAJA {
+namespace util {
+  
+  KokkosPluginLoader::KokkosPluginLoader()
+  {
+    char *env = ::getenv("KOKKOS_PLUGINS");
+    if (env == nullptr)
+    {
+      return;
+    }
+    initDirectory(std::string(env));
+
+    for (auto &func : init_functions)
+    {
+      func(0, kokkos_interface_version, 0, nullptr);
+    }
+  }
+
+  void KokkosPluginLoader::preLaunch(RAJA::util::PluginContext& p)
+  {
+    for (auto &func : pre_functions)
+    {
+      func("", 0, &(p.kID));
+    }
+  }
+
+  void KokkosPluginLoader::postLaunch(RAJA::util::PluginContext& p)
+  {
+    for (auto &func : post_functions)
+    {
+      func(p.kID);
+    }
+  }
+
+  void KokkosPluginLoader::finalize()
+  {
+    for (auto &func : finalize_functions)
+    {
+      func();
+    }
+  }
+
+  // Initialize plugin from a shared object file specified by 'path'.
+  void KokkosPluginLoader::initPlugin(const std::string &path)
+  {
+    void *plugin = dlopen(path.c_str(), RTLD_NOW | RTLD_GLOBAL);
+    if (!plugin)
+    {
+      printf("[KokkosPluginLoader]: dlopen failed: %s\n", dlerror());
+    }
+
+    // Getting and storing init function.
+    init_function init = (init_function) dlsym(plugin, "kokkosp_init_library");
+    if (init)
+      init_functions.push_back(init);
+    else
+      printf("[KokkosPluginLoader]: dlsym failed: %s\n", dlerror());
+
+    // Getting and storing preLaunch function.
+    pre_function pre = (pre_function) dlsym(plugin, "kokkosp_begin_parallel_for");
```

Comment:
@davidbeckingsale should this be changed -- 'parallel_for' to 'forall' -- to make output more familiar to RAJA users?

Reply:
@rhornung67 , if you want to use it with existing Kokkos tools, those tools use this name explicitly. Sorry about that, if we were planning ahead we might have done it differently.

What you might do is ask people to expose another symbol, `begin_forall`, that just calls into the other, I don't know how hard a sell that would be to Tools people though (I could do it for Caliper if it helped)

Reply:
I'm fine keeping it as is. No worries. I was just asking.  'begin_forall' isn't really the best either, since the plugins also work with the RAJA kernel interface.